### DISCUSSION TITLE
the analog sensor does not work in the "counter" mode

### DISCUSSION THREAD

**Discussion by KCH1963** [emsesp/EMS-ESP32#2201] [Q&A]
Water and gas meters are connected according to this diagram.
![image](https://github.com/user-attachments/assets/19eac098-1776-439b-a1f0-8fef403f81fe)
In EMS-ESP it was written like this
![image](https://github.com/user-attachments/assets/3e86693e-0c65-4f2c-8eed-4758d37804d0)
In the Home assistant  I see the meters like this, but they don't count the readings.
![image](https://github.com/user-attachments/assets/f39d365f-2c67-4146-be96-b644e0278f96)
I installed pull-up resistors on the GPIO counters
in the open state of the reed switch at the output of the ESP 32 (GPIO 25/26/27) = 3v3, in the closed state = 0v
signal levels on gpio 25 when water is consumed
![image](https://github.com/user-attachments/assets/a93c36f5-302e-4adf-aa5f-9878c9ac8354)
Soldered a resistor, did not help. In the "counter" mode, the readings are not counted. When setting the GPIO to the "Digital in" mode, the ESP-EMS begins to see pulses and output them to "on / off" No miracle happened
![image](https://github.com/user-attachments/assets/4a6ff947-bfd1-4941-b2cd-0a7c699ac5e9)
It doesn't work with these settings.
![image](https://github.com/user-attachments/assets/6adfa734-742d-40af-aab7-9bae7ffce092)
![Status](https://github.com/user-attachments/assets/55af2b64-9079-419a-b94f-b5becbc308a3)
![Data traffic](https://github.com/user-attachments/assets/0988d35d-eaf2-482f-8d1b-a070f66741e2)

### DESCRIPTION

_A clear and concise description of what the problem is or the change requested._

### REQUESTED INFORMATION

_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Searched the issue in [issues](https://github.com/emsesp/EMS-ESP32/issues)
- [x] Searched the issue in [discussions](https://github.com/emsesp/EMS-ESP32/discussions)
- [x] Searched the issue in the [docs](https://emsesp.org/Troubleshooting/)
- [x] Searched the issue in the [chat](https://discord.gg/3J3GgnzpyT)
- [ ] Provide the System information in the area below, taken from `http://<IP>/api/system`

```json
  Paste System information here....


```

### TO REPRODUCE

_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR

_A clear and concise description of what you expected to happen._

### SCREENSHOTS

_If applicable, add screenshots to help explain your issue._

### ADDITIONAL CONTEXT

_Add any other context about the issue here._

**(Please remember to close the issue when it has been addressed)**

**Comment by MichaelDvP** [comment]
Please add your system info (download from help page). 
Have you tried with different gpio? On esp32 pin 25 and 26 are also dac output, maybe this conflicts with the input-pullup config.

**Reply by KCH1963** [reply]
![Версия](https://github.com/user-attachments/assets/2f907faa-0e11-47d4-884e-39dcfd3c66f6)

**Reply by KCH1963** [reply]
{
  "system": {
    "version": "3.7.1-dev.4",
    "uptime": "000+02:52:33.666",
    "uptimeSec": 10353,
    "platform": "ESP32",
    "cpuType": "ESP32-D0WDQ6",
    "arduino": "Tasmota Arduino v2.0.17",
    "sdk": "4.4.8.240628",
    "freeMem": 167,
    "maxAlloc": 103,
    "freeCaps": 117,
    "usedApp": 1733,
    "freeApp": 251,
    "partition": "app1",
    "resetReason": "RTC watch dog reset: CPU+RTC / APP CPU reset by PRO CPU",
    "psram": false,
    "model": ""
  },
  "network": {
    "network": "WiFi",
    "hostname": "ems-esp",
    "RSSI": -42,
    "TxPowerSetting": 78,
    "staticIP": false,
    "lowBandwidth": false,
    "disableSleep": true,
    "enableMDNS": true,
    "enableCORS": false,
    "APProvisionMode": "disconnected",
    "APSecurity": "wpa2",
    "APSSID": "ems-esp"
  },
  "ntp": {
    "NTPStatus": "connected",
    "enabled": true,
    "server": "time.google.com",
    "tzLabel": "Asia/Yekaterinburg"
  },
  "mqtt": {
    "MQTTStatus": "connected",
    "MQTTPublishes": 3509,
    "MQTTQueued": 0,
    "MQTTPublishFails": 0,
    "MQTTConnects": 1,
    "enabled": true,
    "clientID": "ems-esp",
    "keepAlive": 60,
    "cleanSession": false,
    "entityFormat": 0,
    "base": "ems-esp",
    "discoveryPrefix": "homeassistant",
    "discoveryType": 0,
    "nestedFormat": 1,
    "haEnabled": true,
    "mqttQos": 1,
    "mqttRetain": false,
    "publishTimeHeartbeat": 60,
    "publishTimeBoiler": 0,
    "publishTimeThermostat": 0,
    "publishTimeSolar": 0,
    "publishTimeMixer": 0,
    "publishTimeWater": 10,
    "publishTimeOther": 0,
    "publishTimeSensor": 30,
    "publishSingle": false,
    "publish2command": false,
    "sendResponse": false
  },
  "syslog": {
    "enabled": false
  },
  "sensor": {
    "temperatureSensors": 12,
    "temperatureSensorReads": 24828,
    "temperatureSensorFails": 0,
    "analogSensors": 7,
    "analogSensorReads": 10,
    "analogSensorFails": 0
  },
  "api": {
    "APICalls": 0,
    "APIFails": 0
  },
  "bus": {
    "busStatus": "connected",
    "busProtocol": "Buderus",
    "busTelegramsReceived": 6748,
    "busReads": 1072,
    "busWrites": 0,
    "busIncompleteTelegrams": 10,
    "busReadsFailed": 3,
    "busWritesFailed": 0,
    "busRxLineQuality": 100,
    "busTxLineQuality": 100
  },
  "settings": {
    "boardProfile": "CUSTOM",
    "locale": "en",
    "txMode": 1,
    "emsBusID": 11,
    "showerTimer": false,
    "showerMinDuration": 180,
    "showerAlert": false,
    "phyType": 0,
    "rxGPIO": 23,
    "txGPIO": 5,
    "dallasGPIO": 19,
    "pbuttonGPIO": 0,
    "ledGPIO": 2,
    "hideLed": false,
    "noTokenApi": false,
    "readonlyMode": false,
    "fahrenheit": false,
    "dallasParasite": false,
    "boolFormat": 1,
    "boolDashboard": 1,
    "enumFormat": 1,
    "analogEnabled": true,
    "telnetEnabled": true,
    "maxWebLogBuffer": 25,
    "webLogBuffer": 25,
    "modbusEnabled": false,
    "forceHeatingOff": false,
    "developerMode": false
  },
  "devices": [
    {
      "type": "boiler",
      "name": "CBS/GB*72, Trendline, Cerapur, Greenstar Si",
      "deviceID": "0x08",
      "productID": 123,
      "brand": "Buderus",
      "version": "06.07",
      "entities": 81,
      "handlersReceived": "0x10 0x11 0x15 0x1C 0x18 0x19 0x34 0x2A 0x04",
      "handlersFetched": "0x14 0x16 0x33 0x26 0x28",
      "handlersPending": "0xBF 0xC2 0x1A 0x35 0xD1 0xE3 0xE4 0xE5 0xE6 0xE9 0xEA 0x2E 0x3B"
    },
    {
      "type": "controller",
      "name": "BC25",
      "deviceID": "0x09",
      "productID": 125,
      "brand": "",
      "version": "03.03",
      "entities": 0
    },
    {
      "type": "temperaturesensor",
      "name": "temperaturesensor",
      "entities": 12
    },
    {
      "type": "analogsensor",
      "name": "analogsensor",
      "entities": 7
    },
    {
      "type": "scheduler",
      "name": "scheduler",
      "entities": 2
    }
  ]
}

**Comment by proddy** [comment]
This is from the Discord chat https://discord.com/channels/816637840644505620/816958041345884180/1306610790375100499

I can't reproduce it. I think it's because of the GPIO 25 and dac, as also mentioned in the documentation. https://docs.emsesp.org/Configuring/#adding-analog-and-temperature-sensors

**Reply by KCH1963** [reply]
for this task, GPIO 25, 26 and 27 are used. Write on which GPIO you are interested in the result, I will test them. The pull-up has nothing to do with it, the photo shows that the pulses are formed on the legs of the GPIO and without pull-up resistors

**Reply by KCH1963** [reply]
I will add that in the "Digital in" mode, GPIO 25 and 26 work, but in the "counter" mode they do not work. Doesn't this seem strange to you?

**Reply by proddy** [reply]
yes, this is by design. See the code at https://github.com/emsesp/EMS-ESP32/blob/bf210977452979c8e6390cda9e1a6faf903b64e3/src/analogsensor.cpp#L150-L152

The ESP32 contains two 8-bit DAC (digital to analog converter) channels, connected to GPIO25 (DAC Channel 1) and GPIO26 (DAC Channel 2). Using the DACs allows these channels to be set to arbitrary output voltages between 0...+3.3V (VDD). Both channels can be driven alternatively by an integrated common cosine waveform (CW) generator. Taken from https://github.com/yellobyte/DacESP32

And indeed a pullup isn't needed, as it's just watching 1-0 changes. Would it be easier if you didn't use 25 for the Counter?

**Reply by MichaelDvP** [reply]
Afair the internal pullup of 25/26 is on the DAC  output, so we have to set dac to 255 to have pullup. But maybe setting the dac overwrites line 148: `pinMode(sensor.gpio(), INPUT_PULLUP);`,
We can try to revert the order, set dac first and then the pin to input. 
Strangely for input, timer, rate the dac is not set, it's just INPUT_PULLUP.  
I will search if i have a spare esp32 to test. 
Could you test setting one gpio (e.g. 25) to dac and output 0, then reconfigure to input and check if pullup is ok.

**Reply by MichaelDvP** [reply]
I recalled right, setting gpio25 to digital_out and 0(dac) and afterwards set to digital_in the input is off (low), no pullup.
Setting digital_out to 255, then switch to digital_in set the input to 1. 
Also can confirm that counter mode works if we set dac first and then switch to input. 
fixed in https://github.com/MichaelDvP/EMS-ESP32/releases/tag/latest
But there is an issue with dac value in dashboard (shows only on/off). Will check.

**Reply by proddy** [reply]
thanks @MichaelDvP . After that is confirmed, I can update the documentation.

**Reply by KCH1963** [reply]
Thank you. The meters started working on GPIO21, 22 and 13. You can't even connect relays to GPIO25 and 26, they don't work on these contacts either. The topic can be closed by correcting the documentation on the recommended pins.

**Reply by KCH1963** [reply]
I'll add that the problem arose on ESP 32 DEVKIT v1

**Reply by proddy** [reply]
it's open source. It would be fantastic if you make the PR yourself on the doc to help others.

**Reply by MichaelDvP** [reply]
> The meters started working on GPIO21, 22 and 13.

Counter/input/timer/rate is fixed for gpio 25/26 in my dev. I hoped you test and can confirm this.
> You can't even connect relays to GPIO25 and 26

You have to send 255 for on and 0 for off, these are 8 bit DAC.  
This should also work with relais (i hope you don't try to drive a relais with gpio output).
